ufo2fontir/src/source.rs

```diff
@@ -1236,74 +1259,118 @@ impl Work<Context, WorkId, WorkError> for KerningWork {
-                    Some(KernParticipant::Group((**group_name).clone()))
-                } else {
-                    let glyph_name = GlyphName::from(name);
-                    if !glyph_order.contains(&glyph_name) {
-                        warn!("'{name}' refers to a non-existent glyph; ignored");
-                        return None;
-                    }
-                    Some(KernParticipant::Glyph(glyph_name))
-                }
-            };
+            kerning_groups.locations.insert(pos.clone());
+        }
 
-            for (side1, side2, adjustment) in font.kerning.into_iter().flat_map(|(side1, kerns)| {
-                kerns
-                    .into_iter()
-                    .map(move |(side2, adjustment)| (side1.clone(), side2, adjustment))
-            }) {
-                let (Some(side1), Some(side2)) = (
-                    resolve(&side1, UFO_KERN1_PREFIX),
-                    resolve(&side2, UFO_KERN2_PREFIX),
-                ) else {
-                    warn!(
-                        "{} kerning unable to resolve at least one of '{}', '{}'; ignoring",
-                        source.name.as_ref().unwrap(),
-                        side1.as_str(),
-                        side2.as_str()
-                    );
-                    continue;
-                };
+        context.kerning_groups.set(kerning_groups);
 
-                kerning
-                    .kerns
-                    .entry((side1, side2))
-                    .or_default()
-                    .insert(pos.clone(), (adjustment as f32).into());
+        Ok(())
+    }
+}
+
+/// See <https://github.com/googlefonts/ufo2ft/blob/3e0563814cf541f7d8ca2bb7f6e446328e0e5e76/Lib/ufo2ft/featureWriters/kernFeatureWriter.py#L302-L357>
+impl Work<Context, WorkId, WorkError> for KerningAtWork {
+    fn id(&self) -> WorkId {
+        WorkId::KerningAtLocation(self.at.clone())
+    }
+
+    fn read_access(&self) -> Access<WorkId> {
+        AccessBuilder::new()
+            .variant(WorkId::StaticMetadata)
+            .variant(WorkId::GlyphOrder)
+            .variant(WorkId::KerningGroups)
+            .build()
+    }
+
+    fn exec(&self, context: &Context) -> Result<(), WorkError> {
+        debug!("Kerning for {:#?} at {:?}", self.designspace_file, self.at);
+
+        let designspace_dir = self.designspace_file.parent().unwrap();
+        let static_metadata = context.static_metadata.get();
+        let glyph_order = context.glyph_order.get();
+        let groups = context.kerning_groups.get();
+        let master_locations =
+            master_locations(&static_metadata.all_source_axes, &self.designspace.sources);
+
+        // We know all the groups, read all the kerning and update group naming
+        let source = self
+            .designspace
+            .sources
+            .iter()
+            .find(|s| {
+                !is_glyph_only(s)
+                    && *master_locations.get(s.name.as_ref().unwrap()).unwrap() == self.at
```

Comment:
unrelated, but do we validate at any point that the name always exists? Or would it make sense to store the validated set of master locations in the static metadata?

Reply:
I believe at this point all the named sources have to have locations so this could be made nicer